### PR TITLE
Telemetry request block

### PR DESCRIPTION
**Pull request by matt-bekkers**
Closes the telemetry request issue.

### CODE REVIEW

File: src/include/packet_types.h [hunk 1]
```diff
@@ -150,6 +150,24 @@ void acceleration_data_block_init(AccelerationDataBlock *b, const uint32_t measu
                                   const int8_t full_scale_range, const int16_t x_axis, const int16_t y_axis,
                                   const int16_t z_axis);
 
+typedef struct telemetry_request_block {
+    /**The telemetry request block accessed as a bytes array. */
+    uint8_t bytes[4];
```
Comment (linguini1) [inline src/include/packet_types.h:155]: Please only keep the `bytes[4]` member.

File: src/packet_types.c [hunk 2]
```diff
@@ -193,6 +193,36 @@ void acceleration_data_block_init(AccelerationDataBlock *b, const uint32_t measu
...
+ * */
+void telemetry_request_block_init(TelemetryRequestBlock *b, const uint8_t data_subtype_1,
+                             const uint8_t used_1, const uint8_t data_subtype_2,
+                             const uint8_t used_2, const uint8_t data_subtype_3,
+                             const uint8_t used_3, const uint8_t data_subtype_4,
+                             const uint8_t used_4) { 
+    memcpy(b->bytes, &data_subtype_1, sizeof(data_subtype_1));
+    memcpy(b->bytes + 1 + sizeof(data_subtype_1), &used_1, sizeof(data_subtype_1)); // + 1 and all further incrementations account for the dead space after the data subtype.
+    memcpy(b->bytes + 1 + sizeof(data_subtype_1) + sizeof(used_1), &data_subtype_2, sizeof(data_subtype_2));
+    memcpy(b->bytes + 2 + sizeof(data_subtype_1) + sizeof(used_1) + sizeof(data_subtype_2), &used_2, sizeof(used_2));
```
Comment (linguini1) [inline src/packet_types.c:216]: In this case you can't use `memcpy` for everything. If you take a look at the spec, each byte contains the first 6 bits which is data subtype, a 7th unused bit, and an 8th bit which marks whether the request is valid.

You'll need to use bit shifting. Something like
```c
bytes[0] = (data_subtype_1 << 2);
bytes[0] |= (used_1);
// Repeat for bytes 1, 2 and 3
```

**Review by linguini1 — CHANGES_REQUESTED** [review]

File: src/packet_types.c (hunk 2 above)
Comment (matt-bekkers) [reply src/packet_types.c:216]: When you use the |= operator, I understand that it does a bitwise OR operation, but I have no idea what that means. Documentation online is as readable as egyptian hieroglyphics translated back and forth between random languages on google translate. Is there any way you could ELI5? I understand there are some wacky concepts at play behind that so if you'd rather wait until the Thursday meeting I can do that too.

Comment (matt-bekkers) [reply src/packet_types.c:216]: Thank you! I'm still a little confused but I'll definitely take another crack at it before Thursday.

File: src/packet_types.c [hunk 3]
```diff
@@ -183,16 +183,46 @@ void angular_velocity_block_init(AngularVelocityBlock *b, const uint32_t measure
...
+ * @param data_subtype_3 The third requested data type.
+ * @param used_3 A boolean to determine if the third data request is valid.
+ * @param data_subtype_4 The fourth requested data type.
+ * @param used_4 A boolean to determine if the fourth data request is valid.
+ * */
+void telemetry_request_block_init(TelemetryRequestBlock *b, const uint8_t data_subtype_1, const uint8_t used_1,
+                                  const uint8_t data_subtype_2, const uint8_t used_2, const uint8_t data_subtype_3,
+                                  const uint8_t used_3, const uint8_t data_subtype_4, const uint8_t used_4) {
+
+    b->bytes[0] |= data_subtype_1 << 2;
```
Comment (linguini1) [inline src/packet_types.c:205]: For all the places you use `data_subtype`, please use a regular `=` instead of `|=`.

File: src/packet_types.c [hunk 4]
```diff
@@ -183,16 +183,46 @@ void angular_velocity_block_init(AngularVelocityBlock *b, const uint32_t measure
...
+    b->bytes[0] |= used_1;
+    b->bytes[1] |= data_subtype_2 << 2;
+    b->bytes[1] |= used_2;
+    b->bytes[2] |= data_subtype_3 << 2;
+    b->bytes[2] |= used_3;
```
Comment (linguini1) [inline src/packet_types.c:210]: This is good, but to be really clean, instead of `used_3` do `(used_3 & 0x01)` (this ensures we only OR in the last bit).

Repeat this for all 4 bytes.

File: src/packet_types.c [hunk 5]
```diff
@@ -183,16 +183,35 @@ void angular_velocity_block_init(AngularVelocityBlock *b, const uint32_t measure
...
+void telemetry_request_block_init(TelemetryRequestBlock *b, const uint8_t data_subtype_1, const uint8_t used_1,
+                                  const uint8_t data_subtype_2, const uint8_t used_2, const uint8_t data_subtype_3,
+                                  const uint8_t used_3, const uint8_t data_subtype_4, const uint8_t used_4) {
+
+    b->bytes[0] = data_subtype_1 << 2;
+    b->bytes[0] |= used_1 0x01;
+    b->bytes[1] = data_subtype_2 << 2;
+    b->bytes[1] |= used_2 0x01;
+    b->bytes[2] = data_subtype_3 << 2;
+    b->bytes[2] |= used_3 0x01;
```
Comment (linguini1) [inline src/packet_types.c:210]: Did you try compiling this? It won't compile, you're not using the `&` operator.

**Review by linguini1 — CHANGES_REQUESTED** [review]

**Comment by linguini1** [comment]
Looks great! Can you please re-format everything with clang-format? Once the format check passes I'll merge it.

**Comment by matt-bekkers** [comment]
strange, clang told me everything was OK on my end... I'll try reinstalling clang

**Comment by linguini1** [comment]
> strange, clang told me everything was OK on my end... I'll try reinstalling clang

Still failing, are you sure it's finding the configuration? I also see that you deleted the function for the acceleration data block by accident.

**Review by linguini1 — CHANGES_REQUESTED** [review]
Please remove the two extra files! Also formatting failed again :(

**Review by linguini1 — APPROVED** [review]
Thanks for the patience! Looks awesome!
